Feature/maroon 404 falling ball viscosimeter

danielliegl opened the pull request:
Close #404

danielliegl committed:
initial commit

danielliegl committed:
applying force to the ball

danielliegl committed:
calculation of forces

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
viscosity friction force added

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
Manager and Measurement System

danielliegl committed:
saved

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
start scale & weighable object

danielliegl committed:
fixed buoyancy force and force application a bit

danielliegl committed:
forgot to save scene

danielliegl committed:
ball physics look good

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
Begin Drag and Drop

danielliegl committed:
Drag and Drop and Scale done.

danielliegl committed:
weighable reworked, start pycnometer fill

not working :(

danielliegl committed:
fixed pycnometer fill

danielliegl committed:
begin measuring objects functionality

danielliegl committed:
caliper drag start

danielliegl committed:
added MeasurementManager.cs

danielliegl committed:
forgor

danielliegl committed:
fixed caliper drag and drop for ortho camera

danielliegl committed:
improvements to measurement

danielliegl committed:
calliper improvements

danielliegl committed:
maxspeed debug text, changed floats to decimal for more precision

danielliegl committed:
Merge branch 'feature/Maroon-404_Falling_Ball_Viscosimeter' of https://github.com/GameLabGraz/Maroon into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
moved physics calculation to fixed update

danielliegl committed:
QuantityDecimal implemented

danielliegl committed:
Added Ball Size Slider, Fixed Scale

danielliegl committed:
start of something

danielliegl committed:
fixed pause

danielliegl committed:
physics reworked

danielliegl committed:
ball no longer inside table lmoa

danielliegl committed:
added minSize for Measurement Camera

danielliegl committed:
DragDrop and Measurement improved

danielliegl committed:
better camera controls

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
UI updates and bug fixes

- Ball und Röhre bewegt, damit es besser in schirm passt
- Pyknometer leergewicht bug behoben
- Hahn für Pyknometer
- Messmodus verbessert
- UI Elemente auf die notwendigen reduziert
- Ball Startgröße vergrößert

danielliegl committed:
start of analog caliper

danielliegl committed:
Caliper with Line renderer, probably not gonna be used. Fixed mouse drag when measuring.

danielliegl committed:
Added Better Stopwatch

danielliegl committed:
arg

danielliegl committed:
Merge branch 'feature/Maroon-404_Falling_Ball_Viscosimeter' of https://github.com/GameLabGraz/Maroon into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
Caliper to texture from wikipedia svg

danielliegl committed:
random size for ball and weight, clamp caliper

danielliegl committed:
better pycnometer and scale

danielliegl committed:
added basic oil heating element

danielliegl committed:
Added License for Caliper Sprites

danielliegl committed:
Measuring Tape Display

danielliegl committed:
line renderer for tape

danielliegl committed:
updated physics

danielliegl committed:
added Dropdown for Choosing Fluids

Moved Fluid property calculation to ViscosimeterManager

danielliegl committed:
argh

danielliegl committed:
added more liquids, blur in measurement camera

added Raman

danielliegl committed:
Update Glycerol.asset

danielliegl committed:
new stopwatch, camera

danielliegl committed:
fix build crash

danielliegl committed:
added Simple Tooltip from the Asset Store

danielliegl committed:
First Feedback Changes

- Make Reset Ball more clear
- Make it more clear that slow motion is active.
- Hint for zooming during measurements.
- Change key binding to mouse2 for panning in zoom mode
- disable collider of measured object
- Do not reset temperature after resetting the simulation. Added bool for allowing reset in the QuantityPropertyView

danielliegl committed:
name changes

danielliegl committed:
Experiment added to Maroon Menus + German Translation

danielliegl committed:
fix merge conflict

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.vr.unity:
![grafik](https://github.com/user-attachments/assets/38362d87-7416-4f36-90e5-da856e4b5d5b)
If VR is not supported for the experiment, please delete the scene

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Resources/FallingBallViscosimeter.laboratoryblock.prefab:
Incorrect text in laboratory
![grafik](https://github.com/user-attachments/assets/c49f36f0-6cb1-4578-a58b-1d8f9cc5af03)
(missing translation in Assets/Resources/MLG/laboratory.xml probably)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Resources/FallingBallViscosimeter.laboratoryblock.prefab:
Colliders are too large / too many colliders
![grafik](https://github.com/user-attachments/assets/11b3e83a-fa40-41ee-9fa8-5de56d9ff557)
(only the table and info stand sign thing should have colliders)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Resources/FallingBallViscosimeter.laboratoryblock.prefab:
Please delete unnecessary Coulombs Law stuff from this experiment 
![grafik](https://github.com/user-attachments/assets/a162155f-92a1-46ff-89ca-87629cd8a941)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      //calculate dynamic viscosity
+      decimal dynamic_viscosity = ViscosimeterManager.Instance.calculateDynamicViscosity(temperature);
+      
+      Debug.Log("dynamic visc " + dynamic_viscosity);
```
please remove this Debug.Log

FlorianGlawogger commented on unity/Assets/Maroon/reusableFonts/calling-code/CallingCode-Regular SDF.asset:
Not sure why this file has been changed?
But I don't think anything important changed, so either please roll back to develop for this file, otherwise should also just be fine I suppose

FlorianGlawogger commented on unity/Assets/Maroon/reusableGui/Menu/maroon-reel.renderTexture:
Correct me if I'm wrong, but I don't think you need to change this file? please revert back to develop branch

FlorianGlawogger commented on unity/Assets/Maroon/reusableGui/Experiment/Scripts/Runtime/QuantityPropertyView.cs:
```diff
@@ -153,8 +170,10 @@ private Slider InstantiateSlider(float minValue, float maxValue, Transform paren
-            slider.gameObject.AddComponent<ResetSlider>();
-
+            ResetSlider reset_slider = slider.gameObject.AddComponent<ResetSlider>();
+            reset_slider.AllowReset = allowReset;
```
I think you should also just be able to call
`slider.AllowReset(allowReset);`
instead of those two lines

(But see the comment regarding allowRest in this class, so likely this should just be removed anyways)

FlorianGlawogger commented on unity/Assets/Maroon/reusableGui/Experiment/Scripts/Runtime/QuantityPropertyView.cs:
```diff
@@ -12,6 +12,8 @@ public class QuantityPropertyView : MonoBehaviour
     {
         public QuantityReferenceValue quantity;
 
+        public bool allowReset = true;
```
Is it necessary to add this bool here? Also might lead to confusion in the future since it is only used for the Slider and not the other types of Quantity. I agree that it would be useful to have something like this, but then we'd need to make sure it works for all possibilities.

I just talked with Michael, and we agreed that it's probably the easiest to do it similarly like e.g. `PointWave_Selectionhandler.cs` does it by calling `waveSourceAmplitudeSlider.AllowReset(false);` in the code.

danielliegl replied:
I removed what I added in this class, and did it like the Point Wave experiment.

FlorianGlawogger commented on unity/Assets/Maroon/reusableMaterials/matGlass/Glass.mat:
I think you changed the alpha and blue values of the material? 
please revert back to develop branch for this file, because many other scenes rely on this material, and then those all would look differently and that's not supposed to happen. (If you need your own glass material, you can copy it and place it in the Materials folder of your experiment)

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/AdventureForge/Geometry/Industrial Cabinets and Shelves/MetalCabinet01/Materials/CleanRed/512/MetalCabinetSupportMA512.mat:
I think you maybe enabled Emission by accident? Anyways, please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/AdventureForge/Geometry/Industrial Cabinets and Shelves/MetalCabinet01/Materials/MetalCabientCornersupportMA.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/AdventureForge/Geometry/Industrial Cabinets and Shelves/MetalCabinet01/Materials/MetalCabinetBodyMA.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/planet/Materials/Earth2kMaterial.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/satellite/Materials/aluminum1.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/satellite/Materials/dish_AO_01.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/satellite/Materials/magelan_map1.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/satellite/Materials/magelan_solar.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/tugraz/materials/glass.mat:
changed color of a shared material, Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusablePrefabs/vanDeGraaff/Materials/beige.mat:
Please revert back to develop

FlorianGlawogger commented on unity/Assets/Maroon/reusableGui/Experiment/Scripts/Runtime/QuantityPropertyView.cs:
```diff
@@ -69,6 +71,21 @@ public void ShowUI()
+
+                    sliderDecimal.onValueChanged.AddListener(value =>
+                    {
+                        decimalQuantity.Value = (decimal)value;
```
Is it necessary to have a QuantityDecimal if you use the float value of a Slider anyways? Because if you need higher precision for calculations you could still convert the QuantityFloat to e.g. decimal later in your calculations class. But you know the calculations part of the experiment better than I do, so please just let us know if/why you do need QuantityDecimal ^^
But as far as I can see it, you could use a Float quantity, and then cast it to decimal for your calculations?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      
+      _rigidbody = ball.gameObject.GetComponent<Rigidbody>();
+      fluid_temperature_.minValue = 15.0m;
+      fluid_temperature_.maxValue = 100.0m;
```
I think this should be set in the Inspector 
EDIT: okay, just saw inspector doesn't support decimal property drawers, so if it stays at QuantityDecimal (see comment above) then it's fine, otherwise I'd suggest using in Inspector

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Sprites/ramanheadshot.jpg:
Licensing rights for this image are fine?

danielliegl replied:
Changed the image and added LICENCE.md with license information about the image in the folder

FlorianGlawogger commented on unity/Assets/Maroon/reusableScripts/Tools/StopWatch/StopWatch.cs:
```diff
@@ -50,6 +51,7 @@ public float Elapsed
             }
         }
 
+        public QuantityFloat timeScale = 1;
```
Is this needed? If not please remove?

FlorianGlawogger commented on unity/Assets/Maroon/reusableScripts/Tools/StopWatch/StopWatch.cs:
```diff
@@ -1,5 +1,6 @@
 ﻿using System;
 using System.Collections;
+using Maroon.Physics;
```
If QuantityFloat below is to be removed, then please remove this as well.

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc/LightingData.asset:
Regarding all files in this folder:
Did you force add these? Lighting data should be in .gitignore:
LightingData*
Lightmap*
ReflectionProbe*
*.lighting
*.lighting.meta

Thus I think this whole folder can be removed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.vr/LightingData.asset:
See comment for .pc light data, same applies here as well

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Meshes/pycnometer.fbx:
Great mesh! Is there a reason though why you did not smooth-shade the faces though?
I think it would look even better with smooth shading, you can do it in e.g. Blender by right-clicking on a face in edit mode, and then press shade smooth
![grafik](https://github.com/user-attachments/assets/59d8d2f9-455a-4b74-a130-e7aa3708ee48)
but it's also fine how it looks now IMO

danielliegl replied:
smoothed that mesh :)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Meshes/scale.fbx:
Smooth shading could also be used on the "perimeter sides" of the metal part? :) (highlighted in yellow) But it's also fine how it looks now^^
![grafik](https://github.com/user-attachments/assets/d7955193-186f-49e9-957b-f6d9ec6dbac4)

danielliegl replied:
tried the same as the pycnometer here but got weird visual bugs, so i kept it as is

![grafik](https://github.com/user-attachments/assets/bb843d25-1a70-4786-a67b-68bf2b8951f3)

smooth scale is too powerful for maroon it seems

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Prefabs/Caliper.prefab:
Is the Measuring Tape script needed?
![grafik](https://github.com/user-attachments/assets/c8a93c39-cf1a-4a78-895a-f55bcdf45b3f)
I think it's always disabled, so probably can be removed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Prefabs/Scale.prefab:
Is this scale prefab used? because the scale in the experiment is not this prefab, so I think either this prefab should be deleted, or the actual scale should be this prefab and then used in the scene

danielliegl replied:
removed.

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Prefabs/StopWatchViscosimeter.prefab:
Would be cool if this was a PrefabVariant of the original StopWatch prefab
![grafik](https://github.com/user-attachments/assets/3eca59ad-5e47-4bb6-ae76-a9ec09f800e7)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Prefabs/OilHeatingElement.prefab:
Is this used? It's disabled in the experiment scene, if it's not used it should be deleted

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Resources/FallingBallViscometerExperimentScore.asset:
```diff
@@ -0,0 +1,20 @@
+  m_Script: {fileID: 11500000, guid: 99c98648a2310284996cd0c8309713e1, type: 3}
+  m_Name: FallingBallViscometerExperimentScore
+  m_EditorClassIdentifier: 
+  experimentName: Coulombs Law
```
Wrong name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/FluidViscosityData.cs:
```diff
@@ -0,0 +1,20 @@
+    [CreateAssetMenu(fileName = "Fluid", menuName = "Scriptable Objects/FluidViscosityData", order = 1)]
+    public class FluidViscosityData : ScriptableObject
+    {
+        public string name;
```
hides the name property, 
![grafik](https://github.com/user-attachments/assets/8c90aa83-c730-4e62-a8d2-29f3b5252c33)
so probably should be renamed to something like "fluidName"?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/FluidViscosityData.cs:
```diff
@@ -0,0 +1,20 @@
+        public double density_a;
+        public double density_b;
+
+        public Color fluid_color;
```
I think this is never used? Would definitely be very awesome if it was used though 👍

danielliegl replied:
i think you talk about the color here? was planned, but all liquids were colourless and clear, so i scrapped the idea. removed the color property

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Resources/FluidData/Water.asset:
```diff
@@ -0,0 +1,21 @@
+  m_Script: {fileID: 11500000, guid: 257e23c2724dd59459205f7425ea8cd8, type: 3}
+  m_Name: Water
+  m_EditorClassIdentifier: 
+  name: Water
```
Problem is if German language is used, then it's still called "Water", a solution to this would be to store an MLG key here (rename name property accordingly to e.g. fluidNameTranslationKey or something) and then let the MLG localization translate the name accordingly.

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/CameraControls.cs:
```diff
@@ -0,0 +1,69 @@
+    
+    public float zoomSpeed = 1.1f;
+    
+    // Update is called once per frame
```
default comment can be removed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/CameraControls.cs:
```diff
@@ -0,0 +1,69 @@
+
+//Made using Code from https://www.youtube.com/watch?v=sD0vNXQYY_U
+
+public class CameraControls : MonoBehaviour
```
missing namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/CameraControls.cs:
```diff
@@ -0,0 +1,69 @@
+
+public class CameraControls : MonoBehaviour
+{
+    private float rotationSpeed = 500f;
```
never used

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/CameraControls.cs:
```diff
@@ -0,0 +1,69 @@
+using System.Collections;
+using System.Collections.Generic;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/CameraControls.cs:
```diff
@@ -0,0 +1,69 @@
+    public float zoomSpeed = 1.1f;
+    
+    // Update is called once per frame
+    void Update()
```
method could be made `private` i suppose

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/CameraControls.cs:
```diff
@@ -0,0 +1,69 @@
+        }
+    }
+
+    void Zoom(bool zoomIn)
```
can be made `private`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+      //    transform.localPosition = new Vector3(current_local_position.x, current_local_position.y,
+      //      Mathf.Clamp(transform.localPosition.z, localPositionMin, localPositionMax));
+        //  break;
+      //}
```
please remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+using Maroon.Physics;
+using UnityEngine;
+using UnityEngine.EventSystems;
+using UnityEngine.Purchasing;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+  public delegate void DragEndedDelegate(DragDrop dragDrop);
+  
+  public DragEndedDelegate dragEndedCallback;
+  public bool enabled = true;
```
hides inherited member `enabled`, please use a different name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+using UnityEngine.Purchasing;
+
+
+public class DragDrop : MonoBehaviour
```
missing namespace here as well, please use the namesace e.g. Maroon.Physics.Viscosimeter or something like that for all scripts

danielliegl replied:
i think i moved everything into that namespace now :)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts.meta:
Please add an assembly definition to the FallingBallViscosimeter/Scripts folder (you can look at other experiments like CoulombsLaw to see how they used/named the AssemblyDefinition)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+  public float hoverSizeFactor = 1.2f;
+  public Vector3 planePosition;
+  private Vector3 worldPosition;
+  Plane plane;
```
could be made `private`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
Please also use 4 spaces in this file for indentation, like in the rest of Maroon code

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/DragDrop.cs:
```diff
@@ -0,0 +1,172 @@
+    }
+    if(snap)
+    {
+      dragEndedCallback(this);
```
This is a valid approach to handle snapping, but IMO the name dragEndedCallback suggests that it is always called when dragging ends, even for GameObjects that do not have snap set to true. Thus I would suggest renaming it to something like `dragEndedSnapCallback`?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/SnapController.cs:
```diff
@@ -0,0 +1,47 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class SnapController : MonoBehaviour
```
missing namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/SnapController.cs:
```diff
@@ -0,0 +1,47 @@
+using System.Collections;
```
unnecessary using

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/SnapController.cs:
```diff
@@ -0,0 +1,47 @@
+
+
+
+    // Start is called before the first frame update
```
Please remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/SnapController.cs:
```diff
@@ -0,0 +1,47 @@
+                closestSnapPoint = snapPoint;
+                closestDistance = currentDistance;
+            }
+
+
```
Please remove those 2 empty lines

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/DragDrop/SnapPoint.cs:
```diff
@@ -0,0 +1,8 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class SnapPoint : MonoBehaviour
```
missing namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+using UnityEngine.Purchasing.Extension;
+using UnityEngine.Serialization;
+
+namespace Maroon.Physics
```
Please use a namespace with something like Maroon.Physics.Viscosimeter or something

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Purchasing.Extension;
```
unnecessary using

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+    private Vector3 start_position_;
+    //public decimal start_diameter_ = 1.97m;
+    public decimal start_diameter_ = 30.00m; //millimeter
+    private bool dropped_ = true;
```
I think this variable is never used?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+    private bool dropped_ = true;
+    private bool touching_oil = false;
+    private decimal viscosity_force_ = 0.0m;
+    private decimal buoyancy_force_ = 0.0m;
```
I think these 2 variables are never used?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
please also here use 4 spaces for indentation

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      
+      Mass = ViscosimeterManager.addInaccuracy(Mass);
+      _weighableObject = GetComponent<WeighableObject>();
+      radius_ = start_diameter_ / 2000m;
```
apparently, start_diameter is in mm and radius is in m? What do you think about naming the variable radius_in_m and diameter_in_mm or something similar to avoid future confusion?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+    protected override void HandleUpdate()
+    {
+      
+    }
```
I think unnecessary and could be removed

danielliegl replied:
needs the empty `HandleUpdate` because of it being a `PausableObject`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+
+    protected override void HandleFixedUpdate()
+    {
+      //updateBall();
```
can be removed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+
+
+
+    decimal calculateVolume()
```
could be made private?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      Velocity = 0;
+
+      transform.position = start_position_;
+      //rigidbody_.velocity = Vector3.zero;
```
could be removed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      //rigidbody_.velocity = Vector3.zero;
+    }
+
+    public void dropBall()
```
please start method names with a capital letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+    {
+      if (other.gameObject.layer == 4) //Water
+      {
+        touching_oil = true;
```
if I understand correctly, then the variable `touching_oil`, refers to touching liquid / touching fluid? if so then please rename variable accordingly

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      touching_oil = false;
+    }
+
+    void ApplyFallPhysics()
```
please make `private`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      ApplyForce(f_net);
+    }
+
+    void ApplyForce(decimal force)
```
please make `private`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      }
+    }
+
+    bool CheckCollision(Vector3 position, decimal velocity)
```
please make `private`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Ball.cs:
```diff
@@ -0,0 +1,215 @@
+      return true;
+    }
+
+    float GetPossibleMovement(Ray ray, float range)
```
please make `private`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+using UnityEngine;
+using UnityEngine.UI;
+
+namespace Maroon.Physics
```
see above, use visco something namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+    public bool filled;
+
+    // Start is called before the first frame update
+    void Start()
```
could be made private I suppose

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+    private MeshRenderer renderer_;
+    public bool filled;
+
+    // Start is called before the first frame update
```
can be removed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+    void Update()
+    {
+      
+    }
```
please remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+      
+    }
+
+    public void fillPycnometer() //kg/m^3
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+      filled = true;
+    }
+
+    public void emptyPycnometer()
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
```diff
@@ -0,0 +1,54 @@
+    public void emptyPycnometer()
+    {
+      weighableObject.resetWeight();
+      renderer_.material.color = new Color(0.3f, 0.3f, 0.3f, 0.3f);
```
would be cool if color was the Fluid color (specific to each fluid)

danielliegl replied:
i wanted to do that, but these fluids are all colourless and transparentm so i scrapped that idea

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Pycnometer.cs:
use 4 space indentation

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+using System.Collections.Generic;
+using UnityEngine;
+using TMPro;
+namespace Maroon.Physics
```
namespace
(and please insert blank line above)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
use 4 space indentation

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+using System.Collections;
+using System.Collections.Generic;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+      Gizmos.DrawIcon(on_scale_position, "Light Gizmo.tiff", true);
+    }
+
+    // Start is called before the first frame update
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+    void IResetObject.ResetObject()
+    {
+
+    }
```
I think it should be possible to just write
```
public void ResetObject()
{
}
```

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+      snapPoint = GetComponentInChildren<SnapPoint>();
+    }
+
+    // Update is called once per frame
```
remove comment, could be made private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    // Start is called before the first frame update
+    void Awake()
```
could be made private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/Scale.cs:
```diff
@@ -0,0 +1,57 @@
+      updateText();
+    }
+
+    void updateText()
```
capital starting letter, private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/WeighableObject.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+public class WeighableObject : MonoBehaviour, IResetObject
```
missing namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/WeighableObject.cs:
```diff
@@ -0,0 +1,36 @@
+using System.Collections;
+using System.Collections.Generic;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/WeighableObject.cs:
```diff
@@ -0,0 +1,36 @@
+        weight = new_weight;
+    }
+
+    public void resetWeight()
```
start with capital letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/WeighableObject.cs:
```diff
@@ -0,0 +1,36 @@
+        return weight;
+    }
+
+    public void setWeight(decimal new_weight)
```
start with capital letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/WeighableObject.cs:
```diff
@@ -0,0 +1,36 @@
+        ResetObject();
+    }
+
+    public decimal getWeight()
```
start with capital letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ExperimentObjects/WeighableObject.cs:
```diff
@@ -0,0 +1,36 @@
+    public void setWeight(decimal new_weight)
+    {
+        weight = new_weight;
+    }
```
could be converted into a C# property

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Faucet.cs:
```diff
@@ -0,0 +1,41 @@
+using Maroon.Physics;
+using UnityEngine;
+
+public class Faucet : MonoBehaviour
```
namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Faucet.cs:
```diff
@@ -0,0 +1,41 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
```
remove usings not needed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Faucet.cs:
```diff
@@ -0,0 +1,41 @@
+    void Update()
+    {
+        
+    }
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Faucet.cs:
```diff
@@ -0,0 +1,41 @@
+
+    private void OnMouseDown()
+    {
+        _pycnometer = snapPoint.currentObject.gameObject.GetComponent<Pycnometer>();
```
throws NullReferenceException if snapPoint is empty

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Maroon.Physics
```
namespace naming

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+  public class MeasurableObject : MonoBehaviour
+  {
+    private MeasurementManager _measurementManager;
+    private float length_;
```
don't think these are needed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    // Start is called before the first frame update
+    void Start()
```
remove comment + make private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+    // Start is called before the first frame update
+    void Start()
+    {
+      if(measuredAxis == null)
```
enum cannot be null, the default value is 0 I think, so this if is unnecessary

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
use 4 space indentation

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+      }
+    }
+
+    public void setChooseable(bool chooseable)
```
capital starting letter? 🥺

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+      makeNotChooseable();
+    }
+    
+    private void makeChooseable()
```
capital starting letter? 🥺 🥺

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+      setDragDrop(false);
+    }
+
+    private void makeNotChooseable()
```
capital starting letter? 🥺 🥺  🥺

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+      setDragDrop(true);
+    }
+
+    private void setDragDrop(bool active)
```
capital starting letter? 🥺 🥺 🥺🥺

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+    {
+      clickable = false;
+      setDragDrop(true);
+    }
```
please merge these 3 methods into 1, just have SetChooseable and then just clickable = chooseable; and SetDragDrop(!choosable); or something

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+      if (dragDrop)
+      {
+        dragDrop.enabled = active;
+        Debug.Log(dragDrop.enabled);
```
can this be removed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+    void Update()
+    {
+
+    }
```
please remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurableObject.cs:
```diff
@@ -0,0 +1,102 @@
+
+    private void OnMouseDown()
+    {
+      Debug.Log("Clicked this Boy");
```
Please remove this boy

danielliegl replied:
rip boy

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementBox.cs:
```diff
@@ -0,0 +1,47 @@
+using System.Collections.Generic;
+using UnityEngine;
+
+namespace Maroon.Physics
```
use visco something namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementBox.cs:
```diff
@@ -0,0 +1,47 @@
+using System.Collections;
+using System.Collections.Generic;
```
please remove unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementBox.cs:
```diff
@@ -0,0 +1,47 @@
+      renderer_.enabled = false;
+    }
+
+    void toggleMeasurement(bool mode)
```
capitalize starting letter

danielliegl replied:
removed, was not needed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementBox.cs:
```diff
@@ -0,0 +1,47 @@
+    void Update()
+    {
+    
+    }
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+    public GameObject hintText;
+
+    private bool measuring = false;
+    private float minSize = 0.1f;
```
can these be removed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
indentation

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+using System;
+using System.Collections;
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+using UnityEngine.UI;
+using Object = UnityEngine.Object;
+
+namespace Maroon.Physics
```
namespace naming

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.UI;
+using Object = UnityEngine.Object;
```
can be removed if you just call FindObjectsOfType<MeasurableObject>(); in geAllMeasurableObjects (without the Object.)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+    void Start()
+    {
+    }
+
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+    }
+
+
+    void startMeasuringMode()
```
private + capital

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+      hintText.gameObject.SetActive(true);
+    }
+
+    public void endMeasuringMode()
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+      measurementState = MeasurementState.Off;
+    }
+
+    void fitObjectToCamera()
```
capitalize starting letter, private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+
+    }
+
+    public void chooseMeasuredObject()
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+    }
+
+
+    private void setAllChooseable(bool chooseable)
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+      uiText.SetActive(chooseable);
+    }
+
+    public void setChosenObject(MeasurableObject mObject)
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+    void resetCamera()
+    {
+
+    }
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+
+    }
+
+    void getAllMeasurableObjects()
```
capitalize starting letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+      main_camera.gameObject.tag = "MainCamera";
+      startButton.interactable = true;
+      endButton.interactable = false;
+      measurementState = MeasurementState.Off;
```
could some parts of start and endMeasuringMode methods be grouped into one method that get's passed a parameter? e.g. parts of camera enabling/disabling stuff looks pretty similar

danielliegl replied:
combined into a single method: public void ToggleMeasuringMode(bool isMeasuring)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementManager.cs:
```diff
@@ -0,0 +1,191 @@
+      }
+      
+      
+
```
please remove these 3 empty lines

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/Caliper.cs:
```diff
@@ -0,0 +1,28 @@
+    private TMP_Text text;
+    private float measuredLength;
+    
+    // Start is called before the first frame update
```
remove comment, make method private?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/Caliper.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+    }
+
+
```
please remove these

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/Caliper.cs:
```diff
@@ -0,0 +1,28 @@
+using TMPro;
+using UnityEngine;
+
+public class Caliper : MonoBehaviour
```
missing namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/Caliper.cs:
I think this whole script can be deleted, as no variable/method is used?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/CaliperRenderer.cs:
I think this script is never used and could thus be removed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
```
unnecessary usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+    [SerializeField]private Transform slider;
+    [SerializeField]private Transform head;
+    [SerializeField]private TMP_Text text;
+    [SerializeField]private float measuredLength;
```
please add a space between [SerializeField] and private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+
+    [SerializeField] private Transform lineRendererStart;
+    private Vector3[] positions;
+    [SerializeField]private DragDrop hook_dnd;
```
please add a space between [SerializeField] and private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+    [SerializeField]private Transform head;
+    [SerializeField]private TMP_Text text;
+    [SerializeField]private float measuredLength;
+    [SerializeField] private float offset;
```
please name this a bit more distinctive, what offset?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+    }
+
+    // Update is called once per frame
+    void Update()
```
remove comment, make private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+        updateText();
+    }
+
+    void updateText()
```
make private, capitalize first letter

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+    private void Start()
+    {
+        lineRenderer.positionCount = 2;
+        positions = new Vector3[2];
```
I think this line could be removed if you just write `private Vector3[] positions = new Vector3[2];` when declaring the array

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+        if (transform.rotation.z != 0f)
+        {
+            horizontal = false;
+            offset = -0.03745f;
```
Is it necessary that it's -0.03745f and not -0.0375f ? because if not it could be simplified with the offset if (transform.rotation.z != 0f)

danielliegl replied:
can be the same for both

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/Caliper/MeasuringTape.cs:
```diff
@@ -0,0 +1,65 @@
+    {
+        lineRenderer.positionCount = 2;
+        positions = new Vector3[2];
+        if (transform.rotation.z != 0f)
```
didn't look into this a lot, but are you using the z of the Quaternion on purpose? I.e. usually when writing similar code I check the euler angles z, not the quaternion z

danielliegl replied:
it works with both, didn't put much thought into that. i changed it to euler angles.

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Prefabs/MeasuringTape.prefab:
I cannot move the type when dragging the white center cylinder part of the tape

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/HeaterButton.cs:
```diff
@@ -0,0 +1,59 @@
+{
+    private OilHeatingElement _oilHeatingElement;
+    private MeshRenderer _meshRenderer;
+    private Material _material;
```
can be removed

danielliegl replied:
deleted this script, was unused

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/HeaterButton.cs:
```diff
@@ -0,0 +1,59 @@
+using UnityEngine;
+using UnityEngine.Events;
+
+public class HeaterButton : MonoBehaviour
```
namespace missing

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/HeaterButton.cs:
```diff
@@ -0,0 +1,59 @@
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEngine.Events;
```
remove not required usings

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/HeaterButton.cs:
```diff
@@ -0,0 +1,59 @@
+    [SerializeField] private HeaterButton _oppositeButton;
+    
+    // Start is called before the first frame update
+    void Start()
```
remove comment and make private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/HeaterButton.cs:
```diff
@@ -0,0 +1,59 @@
+    void Update()
+    {
+        
+    }
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/HeaterButton.cs:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    public void turnOff()
```
capital letter start

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+using System.Collections;
+using System.Collections.Generic;
```
remove

danielliegl replied:
deleted this script, was unused

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+using Maroon.Physics;
+using UnityEngine;
+
+public class OilHeatingElement : MonoBehaviour
```
no namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+
+public class OilHeatingElement : MonoBehaviour
+{
+    public static OilHeatingElement Instance;
```
Is it necessary this is a Singleton? Seems not like a usual use case for singleton to me from a quick glance?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+    private ViscosimeterManager _viscosimeterManager;
+    [SerializeField] private bool cooling = false;
+    [SerializeField]private bool heating = false;
+    [SerializeField]private float heatingSpeed = 0.1f;
```
add space after [SerializeField]

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+        Instance = this;
+      }
+    }
+    // Start is called before the first frame update
```
remove comment and make method private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    // Update is called once per frame
+    void Update()
```
remove comment, make method private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Tools/OilHeatingElement/OilHeatingElement.cs:
```diff
@@ -0,0 +1,61 @@
+        }
+
+        return heating;
+    }
```
I'd suggest renaming to ToggleHeating/Cooling, as that name makes it IMO a bit more clear, and using proerties to switch off cooling when heating and the other way around, something like:
```


    public bool Heating
    {
        get { return heating; }
        set { 
            heating = value;
            if (Heating)
                Cooling = false;
        }
    }

    public bool Cooling
    {
        get { return cooling; }
        set
        {
            cooling = value;
            if (Cooling)
                Heating = false;
        }
    }

    public bool ToggleHeating()
    {
        Heating = !Heating;
        return Heating;
    }

    public bool ToggleCooling()
    {
        Cooling = !Cooling;
        return Cooling;
    }
```

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatch.cs:
is this script used? if not please delete

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+using UnityEngine.EventSystems;
+
+[RequireComponent(typeof(EventTrigger))]
+public class StopWatchStartButton : MonoBehaviour
```
namespace

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+{
+    private float hover_timer = 0.0f;
+    private bool hovering = false;
+    [SerializeField]private SimulationStartButton simStartButton;
```
add space after [serializefield]

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+    private TMP_Text text;
+    
+    // Start is called before the first frame update
+    void Start()
```
remove comment and private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    // Update is called once per frame
+    void Update()
```
remove comment and private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+    // Start is called before the first frame update
+    void Start()
+    {
+        text = gameObject.transform.Find("Text (TMP)").GetComponent<TMP_Text>();
```
gameObject.GetComponentInChildren should work

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+    void Start()
+    {
+        text = gameObject.transform.Find("Text (TMP)").GetComponent<TMP_Text>();
+        Debug.Log(text.text);
```
debug.log can be removed probably?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
Once the ball has been dropped, the countdown text still says "3", although it then should be disappear IMO

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+            
+            if (hovering)
+            {
+                hover_timer = hover_timer + Time.deltaTime;
```
use += operator, and I'd suggest using Time.unscaledDeltaTime to avoid waiting very long when using slow mo

danielliegl replied:
ah nice, didn't know that unscaledDeltaTime existed. that fixes that waiting issue thanks

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/Stopwatch/StopWatchStartButton.cs:
```diff
@@ -0,0 +1,64 @@
+                //drop the ball after 3 seconds of hovering
+                hovering = false;
+                hover_timer = 0.0f;
+                simStartButton.StartSimulation();
```
I think you could also just call 
            SimulationController.Instance.StartSimulation();
then you don't need the simStartButton variable

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/Weighable.cs:
can be removed I guess?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Sprites/Materials/Vernier_Caliper_Stick.mat:
I think this would be better placed in FallingBallViscosimeter/Materials

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Sprites/Materials.meta:
if empty, can be removed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+using PlatformControls.PC;
+using UnityEngine;
+using TMPro;
+using Object = UnityEngine.Object;
```
can be removed when using GameObject. instead probably

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+using System.Collections;
+using System.Collections.Generic;
+using Maroon.scenes.experiments.FallingBallViscosimeter.Scripts;
+using PlatformControls.PC;
```
remove unnecessary usings (System.Collections and Platformcontrols.PC)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+using Object = UnityEngine.Object;
+using Random = UnityEngine.Random;
+
+namespace Maroon.Physics
```
use your own more specific namespace for the experiment

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+  public class ViscosimeterManager : MonoBehaviour, IResetObject
+  {
+    public static ViscosimeterManager Instance;
+    public static float inaccuracyRange = 0.1f;
```
only used in this class, doesn't need to be public

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    public Pycnometer pycnometer;
+
+    public decimal fluid_density_ = 0.0m;
+    public QuantityDecimal fluid_temperature_ = 25.0m;
```
like mentioned above, does this need to be decimal?
you could just cast to decimal when using it

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
use 4 space indentation

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      }
+    }
+    
+    MeasurableObject currentMeasurableObject;
```
not needed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    public List<MeasurableObject> measurableObjects;
+
+    private decimal ballMaxSpeed = -1.0m;
+    private Rigidbody _rigidbody;
```
not needed?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      fluid_temperature_.maxValue = 100.0m;
+    }
+
+    private void prepareFluids()
```
capitalize letter starting

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      _current_fluid = _fluids[0];
+      foreach (var fluid in _fluids)
+      {
+        _fluid_dropdown.options.Add(new TMP_Dropdown.OptionData() {text=fluid.name});
```
text should be localized using the LanguageManager

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    }
+    
+    
+    public decimal calculateDynamicViscosity(decimal temperature)
```
please capitalize starting letter of method name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      return (decimal)(_current_fluid.viscosity_a * Math.Pow(_current_fluid.viscosity_b, (double)kelvin) + _current_fluid.viscosity_c);
+    }
+    
+    public decimal celsiusToKelvin(decimal celsius)
```
please capitalize starting letter of method name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    {
+      return celsius + 273.15m;
+    }
+    void calculateFluidDensity()
```
please capitalize starting letter of method name, can this be made private? if not, public

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    }
+    void calculateFluidDensity()
+    {
+      //fluid_density_ = (FluidTemperature * -0.37m + 891.83m); //kg/m^3
```
remove old code comment

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      fluid_density_ = (decimal)celsiusToKelvin(FluidTemperature) * (decimal)_current_fluid.density_a + (decimal)_current_fluid.density_b; //kg/m^3
+    }
+
+    public void togglePycnometerFill(bool fill)
```
please capitalize starting letter of method name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    }
+
+
+    private void toggleMeasurementBoxes(bool mode)
```
please capitalize starting letter of method name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      MeasurementMode = !MeasurementMode;
+    }
+
+    void getAllMeasurableObjects()
```
please capitalize starting letter of method name, private

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      
+    }
+
+    // Update is called once per frame
```
remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    void FixedUpdate()
+    {
+      calculateFluidDensity();
+      //ball.updateBall();
```
remove commented out code

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      update_debug_text();
+    }
+
+    void update_debug_text()
```
please capitalize starting letter of method name, and don't use underscore, use PascalCase

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      ballMaxSpeed = -1;
+    }
+
+    public static float addInaccuracy(float input)
```
please capitalize starting letter of method name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+      return input + Random.Range(-value, value);
+    }
+
+    public static decimal addInaccuracy(decimal input)
```
please capitalize starting letter of method name

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -0,0 +1,185 @@
+    {
+      float value = input * inaccuracyRange;
+      return input + Random.Range(-value, value);
+    }
```
if not needed, please remove

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
Missing Reference: FluidDropDown item
![grafik](https://github.com/user-attachments/assets/c21a9dea-f5a8-4d61-bc67-666719c719e3)
 
(You can use Tools -> MissingReferenceFinder to find missing references, tests require you to have no missing references)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
Tooltips are not localized
![grafik](https://github.com/user-attachments/assets/b8b0c30c-7ccf-495e-9a38-e9f668b7872c)

danielliegl replied:
i wrote a LocalizedSimpleTooltip component. Just need to know where to actually put the code. It's currently in the Scripts/UI folder of my Experiment. This should probably be somewhere, where it makes more sense?

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
When moving the pycnometer out of frame, I can never get it back, it should be returned to the starting pos when resetting simulation IMO

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
Very nitpicky remark, but IMO the picture frames are a bit distracting for the ball camera, 
![grafik](https://github.com/user-attachments/assets/9b0769ba-3d77-499d-be0d-3692c6b180b6)

maybe they could be moved out of the way a bit, like so (though maybe a bit nicer arranged):
![grafik](https://github.com/user-attachments/assets/09e61a8d-4c45-495d-b63e-089c8a28253a)
![grafik](https://github.com/user-attachments/assets/750c4d12-0652-437d-8a04-fb1fd97fc2b6)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
When using any fluid other than 1,2-Ethanediol, (e.g. Cyclohexane), then the ball jumps up and down at the bottom of the cylinder tube
![cursor](https://github.com/user-attachments/assets/ae959b81-9904-4743-855e-9ff7b4f7a0ab)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
not too big of a deal i suppose, but when the ball rotates, the ball camera rotates as well

danielliegl replied:
freezed rotation of the ball

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
When placing the ball, I often placed it to the Faucet snap point by accident
![grafik](https://github.com/user-attachments/assets/c4e9294d-936e-4c47-bc53-0ff22c2861dd)
Then the ball was outside the tube, but it still behaved (falling speed) the same as inside the tube filled with fluid
if it's not too much work it would be cool to prevent the ball from snapping to this snap point

danielliegl replied:
made this snap point only available to the pycnometer

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
Is it on purpose that the tooltip for the tube only appears above and below the red lines? (not in the center)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
When clicking Measure Length and then on e.g. the tube, I get the following error
![grafik](https://github.com/user-attachments/assets/ec1e5fb1-6e32-452d-b9da-b067858e09cb)

FlorianGlawogger reviewed (changes requested):
The experiment is really cool, and very well done! I think it is really well-suited to conduct experiments with it and actually learn about viscosity. 😄 
I made a lot of remarks but mostly about code formatting, naming of methods, missing namespaces etc., so the vast majority of comments will not take up lots of time.
Please also merge with develop branch.
But again: very cool experiment! 👍

danielliegl committed:
Pull Request Changes

danielliegl committed:
Merge branch 'develop' into feature/Maroon-404_Falling_Ball_Viscosimeter

danielliegl committed:
More Pull Request Changes

FlorianGlawogger commented:
Hi, just to be sure we are on the same page @danielliegl , you requested another review, and I saw you made tons of requested changes! Though I looked at a few of the previous review's comments, and it seems like some of the comments are not resolved yet, e.g.  `Debug.Log("dynamic visc " + dynamic_viscosity);` is still in the Ball.cs file, or not sure if you wrote somewhere regarding the Decimal Quantity?
Please tell me in case I am missing something, but otherwise please make sure to resolve all suggestions before requesting another review to save us both some time, (or please write a comment in case a suggestion doesn't make sense).

danielliegl commented:
> Hi, just to be sure we are on the same page @danielliegl , you requested another review, and I saw you made tons of requested changes! Though I looked at a few of the previous review's comments, and it seems like some of the comments are not resolved yet, e.g. `Debug.Log("dynamic visc " + dynamic_viscosity);` is still in the Ball.cs file, or not sure if you wrote somewhere regarding the Decimal Quantity? Please tell me in case I am missing something, but otherwise please make sure to resolve all suggestions before requesting another review to save us both some time, (or please write a comment in case a suggestion doesn't make sense).

Hi @FlorianGlawogger,
I guess I got sloppy after going through these 200 conversations, sorry. I went through all of them again. Now all issues that I have fixed are resolved. There is only the question about the Localization of the Tooltips.

danielliegl committed:
even more pull request changes

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/FallingBallViscosimeter.pc.unity:
When placing the pycnometer flask on the scale, then pressing the reset simulation button, the scale still displays the weight of the pycnometer

danielliegl replied:
fixed

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/MeasurementBox.cs:
please also use 4 space indentation here

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -1,185 +1,168 @@
+using Maroon.UI;
 using UnityEngine;
 using TMPro;
 using Object = UnityEngine.Object;
```
I think you can remove this if you apply the suggestion below for the `GetAllMeasurableObjects` method

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -1,185 +1,168 @@
+
+        private void GetAllMeasurableObjects()
+        {
+            Object[] foundObjects = FindObjectsOfType<MeasurableObject>();
```
```
           MeasurableObject[] foundObjects = FindObjectsOfType<MeasurableObject>();

            measurableObjects = new List<MeasurableObject>();

            foreach (MeasurableObject obj in foundObjects)
            {
                measurableObjects.Add(obj as MeasurableObject);
            }
```
then we can get rid of using Object = UnityEngine.object at the start of the class

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -1,185 +1,168 @@
+            UpdateDebugText();
+        }
+
+        void UpdateDebugText()
```
please make private if possible

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/ViscosimeterManager.cs:
```diff
@@ -1,185 +1,168 @@
+            }
+        }
+
+        void FixedUpdate()
```
please make private if possible

FlorianGlawogger reviewed (changes requested):
Awesome changes 👍  just found one more small bug, and a few nitpicks regarding some code formatting, but I think you will be able to fix those issues very quickly, and then we can finally merge this PR :)

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/LocalizedSimpleTooltip.cs:
```diff
@@ -0,0 +1,23 @@
+using UnityEngine;
+using Vendor.SimpleTooltip;
+
+namespace GEAR.Localization.Text
```
change to `namespace Maroon.UI`

FlorianGlawogger commented on unity/Assets/Maroon/scenes/experiments/FallingBallViscosimeter/Scripts/UI/LocalizedSimpleTooltip.cs:
Please move file to reusableGUI experiment runtime scripts folder 
![grafik](https://github.com/user-attachments/assets/85db62e1-4def-467c-9c14-327d82448bff)

FlorianGlawogger reviewed (changes requested):
just talked with Michael regarding the localized tooltip file, he made a suggestion where to place it.

danielliegl committed:
more PR changes + missing translations

danielliegl commented:
Fixed the last few things. Also fixed another bug, where the "Choose your object to measure" textbox stayed visible, after stopping measurement and added translations for the explanations of the measurement mode.

I hope this should be it :)

FlorianGlawogger reviewed (approved):
Looks good to me! 
Thanks for integrating all the feedback, and for adding a cool new experiment to Maroon 👍 🥳

FlorianGlawogger commented:
fyi: just waiting for Github Actions to finish, then we can merge it
